src/ProgramParser.js

```diff
@@ -1,85 +1,194 @@
 //@flow
 
-import type { Program } from './types';
+import type { Program, ProgramBlock } from './types';
+import { parseLoopLabel } from './Utils';
 
-type ProgramToken = 'forward1' | 'forward2' | 'forward3' |
-                    'backward1' | 'backward2' | 'backward3' |
-                    'left45' | 'left90' | 'left180' |
-                    'right45' | 'right90' | 'right180' |
-                    'eof';
+export type ProgramParserResult = {
+    program: Program,
+    highestLoopNumber: number
+};
 
 export default class ProgramParser {
     text: string;
     textIndex: number;
-    ch: string;
+    ch: ?string;
+    labelsUsed: Set<string>;
+    highestLoopNumber: number;
+    loopLabels: Array<string>;
 
     constructor() {
         this.text = '';
         this.textIndex = 0;
-        this.ch = 'eof';
+        this.ch = null;
+        this.labelsUsed = new Set();
+        this.highestLoopNumber = 0;
+        this.loopLabels = [];
     }
 
-    parse(text: string): Program {
+    parse(text: string): ProgramParserResult {
         this.text = text;
         this.textIndex = 0;
+        this.labelsUsed.clear();
+        this.highestLoopNumber = 0;
         this.nextCh();
+        this.loopLabels = [];
```

Comment:
Please reverse the order of these lines:

```
this.nextCh();
this.loopLabels = [];
```

So that `this.nextCh()` is after we have set all of the other object properties.